### ISSUE TITLE
Is the method of adding noise used in ConstantPerLayerClipper correct?

### ISSUE THREAD

**Issue by Linguanbiao** [pytorch/opacus#202]
Hello,
As we can see , opacus has implemented different clipping strategies, including ConstantPerLayerClipper. This method can specify different layers to use different clipping thresholds.But in the process of using it, I found a problem as shown below, I will show it step by step.

(1)I specify the clipping threshold of all layers in the network to be 1.0.The network has two convolutional layers and two fully connected layers
(2) Clipping layer by layer according to the clipping threshold.
**(3) Add noise, there seems to be a problem here. I specified the clipping threshold of each layer to be 1.0, but Why does the clip_values (max_norm) of each layer become 2.8284, and use this to generate noise?**(As shown in the picture below) 

<img width="800" alt="del" src="https://user-images.githubusercontent.com/43643717/128986139-6b0150f5-e524-4dd5-b0fb-ba74259ac38d.PNG">

## Reason

When I checked the code carefully, I found the cause of the above phenomenon.This is because the following code was written during pre_step. **When returning max_norm to generate noise, the l2 norm is calculated for threshs**, but for the ConstantPerLayerClipper strategy, at this time, threshs contains the clipping threshold of each layer, not a scalar. Therefore, in the above case, tesor([1,1,1,1,1,1,1,1]) will be calculated for the l2 norm, which is 2.8284. But why is it doing this, I think this approach is not feasible for ConstantPerLayerClipper. Similarly, I think new_full() cannot be used on ConstantPerLayerClipper, because the threshold of each layer may be different. The above method is feasible for ConstantFlatClipper because the Threshs of ConstantFlatClipper has only one value.

`       max_norm = threshs.new_full((n,), threshs.norm(2))
        self._reset_aggregated_state()
       `

If the above noise generation method is applied to ConstantPerLayerClipper, more noise will be generated(As shown in the picture below,The clipping threshold of each layer of ConstantPerLayerClipper is 1.0, and the clipping threshold of ConstantFlatClipper is also 1.0,the Dataset is MNIST), so I want to know if this is right, if it should be like this, or if this is a bug in writing, I hope you can answer my question,thanks!
![nochange_2](https://user-images.githubusercontent.com/43643717/128979805-ae8a7b18-d05e-4016-b738-569a4b5bea51.png)

**Comment by alexandresablayrolles** [comment]
Thanks for raising this. In DP-SGD, we add noise proportional to the (maximal) magnitude of the gradient with respect to *all* parameters. As you correctly pointed out, if the gradient of each parameter is clipped to 1, the maximal magnitude of the overall gradient will be bounded by sqrt(8)=2.8284. Hence, the noise is proportional to this.

**Comment by alexandresablayrolles** [comment]
In DP-SGD, we have to add noise to the *whole* parameter gradient and not to individual layer gradients. To do so, we must know the norm of the *whole* gradient. 

A) For ConstantFlatClipper (with e.g. max_grad_norm=1), the norm of the whole gradient is bounded by max_grad_norm and so we add noise proportional to this max_grad_norm (so 1 and not sqrt(8)). 
B) For ConstantPerLayerClipper (with e.g. max_grad_norm=[1.0,0.5, 2.0, 0.2]), the norm of the whole gradient is bounded by sqrt(1.0^2 + 0.5^2 + 2.0^2 + 0.2^2)(this is the Pythagorean theorem). Hence the noise will be scaled according to this value. You are right that it will add "too much" noise to the layer with a max_grad_norm of 0.5, but we have to do this because this is the way DP-SGD works.

**Comment by Linguanbiao** [comment]
@alexandresablayrolles  I understand, thank you very much for your answer！

**Comment by Linguanbiao** [comment]
@alexandresablayrolles  I'm sorry to bother you again. I still have some questions to ask. What do you mean by whole gradient? I noticed that in the implementation process, to find the norm of whole gradient is to first obtain the L2 gradient norm of each sample in each layer, and then calculate the L2 norm again. Is this step necessary? Why is the norm of  whole gradient obtained in this way ? Is it a theoretical requirement? In other words, Why does dpsgd have to satisfy the norm of the whole gradient instead of the norm of a individual layer? Is this the only way for dpsgd? Is there any other way to implement dpsgd that focuses on individual layer gradients? The reason I ask these questions is that I think if we could to focus on individual layer gradients for clipping and adding noise may improve the utility of the model. Thank you!
